### ISSUE TITLE
SPIKE - CDF file generation 

### ISSUE THREAD

**Issue by bryan-harter** [IMAP-Science-Operations-Center/imap_processing#77]
## Description

I'm trying to gather my thoughts here, because I can think of 100 different ways to structure the code and I'm not sure what would be useful for you all and the L3 developers.  These go in order from least hand-holding/overhead to most hand-holding/overhead.  

## Requirements
- Allow instrument data to be written to a CDF file easily, without needing to know much about the internal structure of the CDF file.  

### Nice to have or Goal Requirements
- Checks if ISTP compliant
- Checks if compliant with mission naming standards

## Related tickets
None

## Follow up tickets
This spike is not considered complete until at least one follow up issue is created. 
- [ ] Placeholder ticket

# Solution 1
Simply read instrument data directly into an xarray.DataSet() object.  When you're done manipulating it, call cdflib.xarray.xarray_to_cdf to generate the file.  Keep a dictionary of your attributes for the dataset somewhere in the code.  

## Pros:
- Working directly with xarray is nice, because xarray has some neat data analysis tools that could make processing easier
- More freedom to tweak things in the xarray as you're processing

## Cons:
- You don't know if your xarray/CDF is ISTP compliant until you call the "xarray_to_cdf" function
- You don't know if your xarray meets mission criteria (for example, all variable names must be lower case, etc).  That is only caught outside of the code (code reviews, SPDF reviews, science board meetings, etc)
- No easy way to make a skeleton CDF.  
- Forces us all to use xarray DataSet objects and know how they work
- xarray works best once you have the complete dataset.  In otherwords, you really can't create an empty object and then start appending data to it as you read packets.  It will error because it always wants the size/shape to match between all the variables in the dataset.  So you need to store the data in some intermediate variable, and then read them in all at once to xarray.     

## Additional notes:
- This means there are very few CDF utility functions, other than the one build into cdflib  

# Solution 2
Same as above, but with an extra function living in a utils directory: "verify_imap_cdf()" which loops through the variables in the xarray,DataSet object and ensures they meet certain criteria.  For example:
1) all the names meet mission standards 
2) all the attributes are ISTP compliant 
3) Dimensions of the data match what the attributes say they are
4) etc

## Pros:
- Same as solution 1
- Removes some cons from solution 1

## Cons:
- Forces you to call an extra function before generating the cdfs, what if you forget?!
- Still, xarray works best when you have the complete dataset.  See the Con in Solution 1.  
- Still need to be familiar with xarray.DataSet objects

## Additional notes:
- A CDF utilities module is mainly this 1 function that verifies that the xarray.DataSet object is good to go before converting it to a CDF.  

# Solution 3
Create a "Template" python class for each type of data file.  For example: https://github.com/bryan-harter/imap_processing/blob/dev/imap_processing/idex/IDEXL1aTemplate.py

This is just a not-so fancy wrapper around a python dictionary object.  This dictionary will be set up in the way xarray likes (see https://docs.xarray.dev/en/stable/generated/xarray.DataArray.from_dict.html).  The initial variables, global attributes, variable attributes and dimensions must be set by hand in __init_().  When it comes time to insert data, you'll call the "insert_data" function, which performs a few checks before allowing data to be inserted into the dictionary.  

Once the data is fully inserted into the dictionary object, you can make an xarray.DataSet object from it by calling {template}.to_xarray(), or CDF files by calling {template}.to_cdf().  

## Pros:
- All attributes are centrally located in the __init__() function
- Can make a skeleton cdf object by calling {template}.to_cdf() before you populate with data 
- Can check data when its inserted into dictionary object 
- You can work with the data however you like before you insert it into the template
- Can still get an xarray object out to manipulate for further processing.  

## Cons:
- Data isn't being read into an xarray object until {template}.to_xarray() is called.
- You're not really sure if your initial attributes are correct until you call {template}.to_cdf()
- If you pull an xarray object out of the template to perform data processing (from l1a->l1b for example), it's kind of awkward to force ourselves to read it back into a L1bTemplate object.  In theory the processed dataset object is good to go, and you're immediately able to call the "xarray_to_cdf" function.  *But*, if we don't include the TemplateL1b class, there's no easy way to generate a skeleton CDF from it.  

## Additional notes:
- Moderate amounts of hand-holding
- There's not like a whole library of utility tools that we'd use here, this would be more of a pattern of implementation.  Each of us would be responsible for defining our own template class, albeit with a lot of ability to copy+paste.  There would probably be a small base class that we'd inherit from.  

# Solution 4
A Mix of 3 and 2.  Have a function inside the Template class that verifies the global/variable attributes at the end of the __init__() function.  

## Pros:
- Verifies that the attributes are correct when you make a new instance of the class

## Cons:
-Kind of weird to have a class verify itself?  

## Additional notes:

# Solution 5
Solution 3, but with a twist: instead of defining a bunch of dictionary object in __init__(), we have a "CDFBuilder" class that builds up the dict object from solution 4 while actively checking to ensure all the attributes look correct.  For example, we could have code that looks like:

```
idexl1a_builder = CDFBuilder()
idexl1a_builder.new_coordinate(name='Epoch', dimensions=('Epoch'), attributes=idex_l1a_epoch_attributes)
idexl1a_builder.new_variable(name='TOF_Low', dimensions=('Epoch'), attributes=idex_l1a_tof_low_attributes)

idexl1a_template = CDFTemplate(idexl1a_builder)
idexl1a_template.insert_coordinate_data('Epoch', [1,2,3,4])
idexl1a_template.insert_variable_data('TOF_Low', [1,2,3,4])
idexl1a_template.to_cdf()
```

## Pros:
- Ensures the data is compliant as you're creating the data object, as opposed to when you go to make the CDF/xarray
- Could be useful for the Level 3 developers if they desire more hand holding for CDF file creation, especially if we aren't reviewing their code

## Cons:
- Would require quite a bit of up-front time coding and documenting how the CDFBuilder and the CDFTemplate classes work.  
- Is kind of heavy-handed when it comes to CDF file creation
- The code doesn't look like a nice neat dictionary object that, imo is easier to read than calling a bunch of functions 
- In theory our attributes will be caught via code reviews, science control board meetings, and iterating with the SPDF, so these additional verification steps might not buy us much? (especially if the verification is done at the time of CDF file creation too).  
- Still has the con that you're not working with xarray directly

## Additional notes:

# Summary
There's like 100 ways to do this.  I'm 100% fine implementing whatever! I just didn't want to go deep into one of the solutions without a discussion about the pros and cons first. 

I know what's easiest for IDEX might not be easiest for the other instruments.  And part of the problem is, we might not know what we want until we get deep into algorithm development (for example, exactly how useful will it be for us to process data within the xarray.DataSet objects?)

**Comment by maxinelasp** [comment]
I like solution 3 or 4 because I feel like it maximizes the amount of shared code / implementation work between instruments. It is more or less what I was picturing during our discussion yesterday. We could share sections of the code as needed as well - if there are similar attributes between two instruments, that could be a shared class which goes into the Lo template or whatever. I like the idea of having multiple small dictionaries which combine into the full template which combines into the final CDF. 

It seems also like we should decide if this is going to be primarily internal vs primarily external - personally, I think solution 5 is overkill for an internal only tool. From my view, I don't think most people will want to use this, and those that would can always use the more involved version in Solutions 1-4. Since this is more of a handy tool rather than something to meet a requirement, I think it's not necessary to do much hand-holding.

**Comment by tech3371** [comment]
I agree. It's hard to know what we need exactly until we get deeper into algorithm implementation. I think if we start with what we know, then add as we know more is a good approach. For example, we know global project variables will be shared with all instruments. If we have `CDFBuilder()` class with that global variable in it, I feel like it's a good start. That class could have common methods as you see fit. Eg. `to_cdf()`, `to_xarray()`, `verify_imap_cdf()` (I don't know if they fit in that class or what they do in details, but I am listing them as an example)

And I agree with Maxine. As we continue, if people see that some instruments can share code or implementation, we can refactor to add that.

**Comment by greglucas** [comment]
I think one of the things to think about is when you want to add the metadata. You can create the skeletons upfront and then fill in data (metadata first), or you could create all your data and then assemble the dataset and add the metadata (metadata last).

Also, just for reference, here are the common climate and forecasting metadata conventions for reference in case someone wants to compare against ISTP.
CF: https://cfconventions.org/cf-conventions/cf-conventions.html
ISTP: https://spdf.gsfc.nasa.gov/istp_guide/istp_guide.html

@bryan-harter I like that you're thinking about how to verify this too. I actually think we should make it a requirement to add an ISTP checker as a unit test for every file getting produced. That way our unit tests would catch any changes if someone renames something or changes/adds an attribute that doesn't conform.

My two cents on the suggestions are very similar to what others have already written which is a good thing we are all basically on the same page :) 
I like the suggestion of having standardized attribute dictionaries that could be combined/re-used where appropriate! I think this would be similar to what you've already implemented in solution 3 except that I probably wouldn't go the class route as I feel that adds a bit of indirection versus just giving some good examples of how to make dataarrays and datasets. (Or perhaps make a helper function instead of an entire class that you need to learn new methods for)

```python
MISSION_ATTRS = {"mission_name": "IMAP", ...}
LO_ATTRS = {"instrument": "IMAP_Lo", ...}
VARIABLE_ATTRS = {"units": "#/s", ...}

# single data array
da = xr.DataArray(..., attrs=VARIABLE_ATTRS)
# combine multiple data arrays and align them properly
ds = xr.DataSet([da, ...], attrs=MISSION_ATTRS | LO_ATTRS)
# export to CDF
cdflib.xarray.xarray_to_cdf(ds)
```

**Comment by sdhoyt** [comment]
For the number 4 verification, is that a verification of what already exists in the class or would that verify data added to the object instance of the template class?

**Comment by GFMoraga** [comment]
I'm kind of into solution 4 due to the hybrid nature and proficiency. Having a class double-check itself might seem a bit unusual because there could be biases, but also it could serve as an auto-proofreader.  How well does Solution 4 scale as the complexity of the codebase and the number of attributes increase? Does it remain effective in handling a larger scope?

**Comment by bryan-harter** [comment]
I'll respond over the span of a few comments if ya'll don't mind!

@greglucas That's interesting, so what you're kinda suggesting is that we don't do any checking of the CDF file while the processing code is running, we just make these xarray objects (with reusable dictionaries) and run "xarray_to_cdf" directly.  **But** we make it a requirement that all new CDF files must have a unit tests, and the unit tests must check for ISTP and IMAP mission specific compliance?  I'm good with that as well.  It would certainly leave less to mess around with in the code but still pretty much ensuring the files are compliant.

The only reason I'd be hesitant about that is if somewhere along in the data processing, the xarray object gets changed in an unexpected way.  Like maybe the data we receive is slightly different than our unit test data, which causes the output from our processing to be different from usual? 

The xarray_to_cdf function actually checks for ISTP compliance, but it doesn't throw an error, it just prints out that the file isn't ISTP compliant.  One of the things I could implement there is a flag that says "force_istp" or something, and the function throws and error if there are any irregularities.  As in, this would be a change *within* cdflib.xarray.xarray_to_cdf.  This wouldn't solve any sort of checks we want to do that are mission specific though (for example, ensure all variable names start with "imap", all variable names are lower-case, etc)

**Comment by bryan-harter** [comment]
@maxinelasp I figured this would primarily be just us, perhaps with some of the level 3 developers using it.  So really, just an internal tool.  I definitely wasn't planning on it being the latest and greatest piece of CDF technology.  It probably isn't necessary, but it could be nice if a few years from now the SDC is being supported by entirely new people and they have those functions to use.

**Comment by bryan-harter** [comment]
@sdhoyt What I was thinking (very preliminarily) was that the function would check the initial dictionary attributes created in the __init__() function.  So if there was a base class that all CDF classes inherited from, that base class's wouldn't let inherited objects be instantiated if their attributes were incorrect.  

It is very possible this is all overkill though for tools that are just for us, especially if we unit test the file to ensure compliance

**Comment by greglucas** [comment]
> The only reason I'd be hesitant about that is if somewhere along in the data processing, the xarray object gets changed in an unexpected way. Like maybe the data we receive is slightly different than our unit test data, which causes the output from our processing to be different from usual?

This makes sense to me as well. I like the idea of making sure we are testing for ISTP compliance right away within unit tests (my comment), but then also as you're suggesting adding a runtime check to make sure we didn't miss anything or something else didn't slip in. So 👍 from me.

**Comment by maxinelasp** [comment]
Related PR: #179

**Comment by sdhoyt** [comment]
SPIKE closed by issue #125
